src/expression.d

```diff
@@ -9467,6 +9467,19 @@ extern (C++) final class DelegateExp : UnaExp
         AggregateDeclaration ad = f.toParent().isAggregateDeclaration();
         if (f.needThis())
             e1 = getRightThis(loc, sc, ad, e1, f);
+        if (f.type.ty == Tfunction)
+        {
+            TypeFunction tf = cast(TypeFunction)f.type;
+            if (!MODimplicitConv(e1.type.mod, f.type.mod))
+            {
+                OutBuffer thisBuf, funcBuf;
+                MODMatchToBuffer(&thisBuf, e1.type.mod, tf.mod);
+                MODMatchToBuffer(&funcBuf, tf.mod, e1.type.mod);
+                error("%smethod %s is not callable using a %s%s",
+                    funcBuf.peekString(), f.toPrettyChars(), thisBuf.peekString(), e1.toChars());
```

Comment:
Hm you may be right. I was under the wrong impression that we used to print the type name, but existing diagnostics are:

```D
class C
{
    void ping() { }
}

void main()
{
    auto c = new const C;
    c.ping();
}
```

```
Error: mutable method test.C.ping is not callable using a const object
```

I suppose your approach works fine. :)